File: skeleton/src/main/java/me/shkschneider/skeleton/ui/MyScrollView.java
package me.shkschneider.skeleton.ui;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.util.AttributeSet;
import android.widget.ScrollView;

import me.shkschneider.skeleton.helper.AndroidHelper;

// <http://cyrilmottier.com/2013/05/24/pushing-the-actionbar-to-the-next-level/>
public class MyScrollView extends ScrollView {

    private OnScrollViewListener mOnScrollViewListener;

    public MyScrollView(final Context context) {
        super(context);
    }

    public MyScrollView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
    }

    public MyScrollView(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(AndroidHelper.API_21)
    @SuppressWarnings("NewApi")
    @SuppressLint("NewApi")
    public MyScrollView(final Context context, final AttributeSet attrs, final int defStyleAttr, final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public void setOnScrollViewListener(final OnScrollViewListener onScrollViewListener) {
        mOnScrollViewListener = onScrollViewListener;
    }

    @Override
    protected void onScrollChanged(final int l, final int t, final int oldl, final int oldt) {
        super.onScrollChanged(l, t, oldl, oldt);
        if (mOnScrollViewListener != null) {
            mOnScrollViewListener.onScrollChanged(l, t, oldl, oldt);
        }
    }

    public interface OnScrollViewListener {

        public void onScrollChanged(int l, int t, int oldl, int oldt);

    }

}

File: core/src/main/kotlin/me/shkschneider/skeleton/ui/widget/MyScrollView.kt
package me.shkschneider.skeleton.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ScrollView

import me.shkschneider.skeleton.helperx.log.Logger

private const val PARALLAX = 0.1.toFloat()

// <http://cyrilmottier.com/2013/05/24/pushing-the-actionbar-to-the-next-level/>
// <https://stackoverflow.com/a/26990539/603270>
// MyScrollView.setOnScrollViewListener()
class MyScrollView(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : ScrollView(context, attrs, defStyleAttr) {

    private var onScrollViewListener: OnScrollViewListener? = null
    private var view: View? = null
    private var parallax: Float? = null

    fun setOnScrollViewListener(onScrollViewListener: OnScrollViewListener) {
        this.onScrollViewListener = onScrollViewListener
    }

    override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
        super.onScrollChanged(l, t, oldl, oldt)
        parallax?.let { parallax ->
            view?.translationY = scrollY * parallax
        }
        onScrollViewListener?.onScrollChanged(l, t, oldl, oldt)
    }

    fun parallax(view: View, parallax: Float = PARALLAX) {
        this.view = view
        this.parallax = parallax
        if (parallax <= 0) {
            Logger.warning("Parallax effect too low")
            this.parallax = PARALLAX
        }
        if (parallax > 1) {
            Logger.warning("Parallax effect too high")
            this.parallax = PARALLAX
        }
    }

    interface OnScrollViewListener {

        fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int)

    }

    companion object {

        fun canScroll(scrollView: ScrollView): Boolean {
            val childHeight = scrollView.height
            return scrollView.height < childHeight + scrollView.paddingTop + scrollView.paddingBottom
        }

    }

}
